### PR TITLE
Feature/nex 1238/test preview init

### PR DESCRIPTION
**Pull request by boajer**
https://oat-sa.atlassian.net/browse/NEX-1238
created the endpoint / controller for Fe,  and included FE development inside this

### CODE REVIEW

**Comment by gabrielfs7** [comment]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

File: controller/TestPreviewer.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
+<?php
+
+declare(strict_types=1);
```
Comment (gabrielfs7) [inline controller/TestPreviewer.php:3]: Should be after the doc, no?

File: doc/taoQtiTestPreviewer.yaml [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+  responses:
+    init:
+      description: Good response
+      content:
+        application/json:
+          schema:
+            $ref: '#/components/schemas/testInit'
+
+    badResponse:
```
Comment (gabrielfs7) [inline doc/taoQtiTestPreviewer.yaml:26]: IMO we should use components and schemas with first upper case (like Php class name) and only attributes with lower case. We are doing this to all other api docs.

Comment (gabrielfs7) [inline doc/taoQtiTestPreviewer.yaml:26]:
```suggestion
    BadResponse:
```

And do the same for the others.

File: controller/TestPreviewer.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            $this->setSuccessJsonResponse(
+                [
+                    'success' => true,
+                    'testData' => [],
+                    'testContext' => [],
+                    'testMap' => $response->getMap()->getMap(),
+                ]
+            );
+        } catch (Throwable $exception) {
+            $message = ($exception instanceof common_exception_UserReadableException) ? $exception->getUserMessage(
```
Comment (gabrielfs7) [inline controller/TestPreviewer.php:70]:
```suggestion
            $message = $exception instanceof common_exception_UserReadableException ? $exception->getUserMessage(
```

File: controller/TestPreviewer.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+class TestPreviewer extends tao_actions_ServiceModule
+{
+    use HttpJsonResponseTrait;
+
+    /**
+     * @security("hide")
+     */
+    public function __construct()
+    {
+        taoQtiTest_helpers_TestRunnerUtils::noHttpClientCache();
```
Comment (gabrielfs7) [inline controller/TestPreviewer.php:43]: You should in the `init()` method instead, cause the headers are sent based on response object.

```
$this->getResponseFormatter()
                ->addHeader('...');
``

File: controller/TestPreviewer.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                [
+                    'success' => true,
+                    'testData' => [],
+                    'testContext' => [],
+                    'testMap' => $response->getMap()->getMap(),
+                ]
+            );
+        } catch (Throwable $exception) {
+            $message = $exception instanceof common_exception_UserReadableException ? $exception->getUserMessage(
+            ) : $exception->getMessage();
```
Comment (bartlomiejmarszal) [inline controller/TestPreviewer.php:66]: This is more readable?

```suggestion
            $message = $exception instanceof common_exception_UserReadableException 
            ? $exception->getUserMessage() 
            : $exception->getMessage();
```

File: doc/taoQtiTestPreviewer.yaml [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+            type: string
+          required: true
+          example: 'https%3A%2F%2Fcommunity.docker.localhost%2Fontologies%2Ftao.rdf%23i5efb42fced1ce7064796c08b38ee176'
+          description: Url encoded test identifier
+
+      summary: Returns a test map
+      responses:
+        '200':
+          $ref: '#/components/responses/InitResponse'
+        '500':
```
Comment (mike-ionut-mihai-sandu-tao) [inline doc/taoQtiTestPreviewer.yaml:29]: either this is wrong or the line 223

File: controller/TestPreviewer.php (hunk 5 above)
Comment (gabrielfs7) [reply controller/TestPreviewer.php]: Agree and added extra line also

File: doc/taoQtiTestPreviewer.yaml (hunk 6 above)
Comment (mike-ionut-mihai-sandu-tao) [reply doc/taoQtiTestPreviewer.yaml:29]: are we providing just 200 and 500 ? no 400 about a bad request?

File: doc/taoQtiTestPreviewer.yaml [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+paths:
+  /taoQtiTestPreviewer/TestPreviewer/init:
+    get:
+      parameters:
+        - in: query
+          name: testUri
+          schema:
+            type: string
+          required: true
+          example: 'https%3A%2F%2Fcommunity.docker.localhost%2Fontologies%2Ftao.rdf%23i5efb42fced1ce7064796c08b38ee176'
```
Comment (mike-ionut-mihai-sandu-tao) [inline doc/taoQtiTestPreviewer.yaml:22]: localhost?

File: doc/taoQtiTestPreviewer.yaml (hunk 6 above)
Comment (gabrielfs7) [reply doc/taoQtiTestPreviewer.yaml]: Line 223 is just `example: 400`, but yes. The server sends a 400 :) 

Fixed

File: controller/TestPreviewer.php [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            $requestParams = $this->getPsrRequest()->getQueryParams();
+
+            if (empty($requestParams['testUri'])) {
+                throw  new InvalidArgumentException('Required `testUri` param is missing ');
+            }
+
+            /** @var TestPreviewerService $previewer */
+            $previewer = $this->getServiceLocator()->get(TestPreviewerService::class);
+
+            $response = $previewer->createPreview(new TestPreviewRequest($requestParams['testUri']));
```
Comment (mike-ionut-mihai-sandu-tao) [inline controller/TestPreviewer.php:49]: maybe this should have been in a  private method

**Comment by mike-ionut-mihai-sandu-tao** [comment]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful

File: doc/taoQtiTestPreviewer.yaml (hunk 7 above)
Comment (gabrielfs7) [reply doc/taoQtiTestPreviewer.yaml:22]: Yes, all places are using localhost. What do you suggest?

File: controller/TestPreviewer.php (hunk 8 above)
Comment (gabrielfs7) [reply controller/TestPreviewer.php]: agree. Moved

**Pull request merged**